panic: index out of range in GetMainText across 8 providers (same bug as #422)

## Problem
`GetMainText` panics on any streamed line lacking `"data: "` prefix (e.g. SSE keep-alive comments like `": ping"`). Same class of bug as #121 and #422 (closed as pollinations-only). Present in 8 providers.

## Steps
Any provider whose upstream SSE stream sends a non-`data:` line before a `data:` line will crash. Reproduces logically:
```go
line := ": ping"                   // valid SSE keep-alive, len > 1
strings.Split(line, "data: ")[1]   // panic: index out of range [1] with length 1
```
The guard `if len(line) > 1` does not check for the substring.

## Expected
Return `""` when `"data: "` prefix is absent, matching the JSON unmarshal error branch.

## Actual
```
panic: runtime error: index out of range [1] with length 1
```

## Affected files (all identical pattern)
- `src/providers/deepseek/deepseek.go:87`
- `src/providers/gemini/gemini.go:95`
- `src/providers/groq/groq.go:88`
- `src/providers/kimi/kimi.go:141`
- `src/providers/minimax/minimax.go:88`
- `src/providers/openai/openai.go:107`
- `src/providers/pollinations/pollinations.go:102`
- `src/providers/sky/sky.go:75`

Suggested fix: `if !strings.Contains(line, "data: ") { return "" }` before splitting, or use `strings.SplitN` + length check.

## Environment
tgpt main @ 75a182a, Go 1.24 (go.mod).

Thanks for maintaining aandrew-me/tgpt!

--- a/.gitignore
+++ b/.gitignore
@@ -9,3 +9,4 @@ CLAUDE.md
 
 # Tools directory - development/testing utilities
 tools/
+.opencode/

--- a/md/providers.md
+++ b/md/providers.md
@@ -7,6 +7,7 @@
 - [MiniMax](https://platform.minimaxi.com/) (Requires API key, default model `MiniMax-M2.7`)
 - [Ollama](https://www.ollama.com/) (Local models) (Supports many models)
 - [OpenAI](https://platform.openai.com/docs/guides/text-generation/chat-completions-api) (All models, Requires API Key, supports custom endpoints)
+- [OpenCode](https://opencode.ai) (Free, uses opencode.ai/zen API, default model deepseek-v4-flash-free, default API key `public`, supports OPENCODE_API_KEY / OPENCODE_MODEL / OPENCODE_URL env vars)
 - [Pollinations](https://pollinations.ai/) ([Many free models](https://text.pollinations.ai/models))
 - [Gemini](https://gemini.google.com) (Require a free API keys, supports [many models](https://ai.google.dev/gemini-api/docs/models/gemini), default model `gemini-2.0-flash`)
 

--- a/src/helper/helper.go
+++ b/src/helper/helper.go
@@ -872,7 +872,7 @@ func ShowHelpMessage() {
 
 	boldBlue.Println("\nProviders:")
 	fmt.Println("The default provider is pollinations. The AI_PROVIDER environment variable can be used to specify a different provider.")
-	fmt.Println("Available providers to use: deepseek, gemini, groq, isou, koboldai, minimax, ollama, openai, sky and pollinations")
+	fmt.Println("Available providers to use: deepseek, gemini, groq, isou, koboldai, minimax, ollama, openai, opencode, pollinations, powerbrain and sky")
 
 	bold.Println("\nProvider: deepseek")
 	fmt.Println("Uses deepseek-reasoner model by default. Requires API key. Recognizes the DEEPSEEK_API_KEY and DEEPSEEK_MODEL environment variables")
@@ -894,6 +894,9 @@ func ShowHelpMessage() {
 
 	bold.Println("\nProvider: ollama")
 	fmt.Println("Needs to be run locally. Supports many models")
+
+	bold.Println("\nProvider: opencode")
+	fmt.Println("Free provider using opencode.ai/zen API. Uses deepseek-v4-flash-free model by default. API key defaults to 'public'. Recognizes the OPENCODE_API_KEY, OPENCODE_MODEL and OPENCODE_URL environment variables.")
 
 	bold.Println("\nProvider: openai")
 	fmt.Println("Needs API key to work and supports various models. Recognizes the OPENAI_API_KEY and OPENAI_MODEL environment variables. Supports custom urls with --url")

--- a/src/helper/helper.go
+++ b/src/helper/helper.go
@@ -872,7 +872,7 @@ func ShowHelpMessage() {
 
 	boldBlue.Println("\nProviders:")
 	fmt.Println("The default provider is pollinations. The AI_PROVIDER environment variable can be used to specify a different provider.")
-	fmt.Println("Available providers to use: deepseek, gemini, groq, isou, koboldai, minimax, ollama, openai, opencode, pollinations, powerbrain and sky")
+	fmt.Println("Available providers to use: deepseek, gemini, groq, isou, kimi, koboldai, minimax, ollama, openai, opencode, pollinations, powerbrain and sky")
 
 	bold.Println("\nProvider: deepseek")
 	fmt.Println("Uses deepseek-reasoner model by default. Requires API key. Recognizes the DEEPSEEK_API_KEY and DEEPSEEK_MODEL environment variables")

--- a/src/providers/deepseek/deepseek.go
+++ b/src/providers/deepseek/deepseek.go
@@ -83,7 +83,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/gemini/gemini.go
+++ b/src/providers/gemini/gemini.go
@@ -91,7 +91,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 }
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/groq/groq.go
+++ b/src/providers/groq/groq.go
@@ -84,7 +84,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/kimi/kimi.go
+++ b/src/providers/kimi/kimi.go
@@ -137,7 +137,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/minimax/minimax.go
+++ b/src/providers/minimax/minimax.go
@@ -84,7 +84,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/openai/openai.go
+++ b/src/providers/openai/openai.go
@@ -103,7 +103,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/opencode/opencode.go
+++ b/src/providers/opencode/opencode.go
@@ -1,0 +1,112 @@
+package opencode
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+
+	http "github.com/bogdanfinn/fhttp"
+
+	"github.com/aandrew-me/tgpt/v2/src/client"
+	"github.com/aandrew-me/tgpt/v2/src/structs"
+)
+
+type RequestBody struct {
+	Model    string `json:"model"`
+	Stream   bool   `json:"stream"`
+	Messages []any  `json:"messages"`
+}
+
+func NewRequest(input string, params structs.Params) (*http.Response, error) {
+	client, err := client.NewClient()
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(0)
+	}
+
+	model := "deepseek-v4-flash-free"
+	if params.ApiModel != "" {
+		model = params.ApiModel
+	} else if envModel := os.Getenv("OPENCODE_MODEL"); envModel != "" {
+		model = envModel
+	}
+
+	apiKey := "public"
+	if params.ApiKey != "" {
+		apiKey = params.ApiKey
+	} else if envKey := os.Getenv("OPENCODE_API_KEY"); envKey != "" {
+		apiKey = envKey
+	}
+
+	url := params.Url
+	if url == "" {
+		if envUrl := os.Getenv("OPENCODE_URL"); envUrl != "" {
+			url = envUrl + "/v1/chat/completions"
+		}
+	}
+
+	if url == "" {
+		url = "https://opencode.ai/zen/v1/chat/completions"
+	}
+
+	requestInfo := RequestBody{
+		Model:  model,
+		Stream: true,
+		Messages: []any{
+			structs.DefaultMessage{
+				Content: params.SystemPrompt,
+				Role:    "system",
+			},
+		},
+	}
+
+	if len(params.PrevMessages) > 0 {
+		requestInfo.Messages = append(requestInfo.Messages, params.PrevMessages...)
+	}
+
+	requestInfo.Messages = append(requestInfo.Messages, structs.DefaultMessage{
+		Role:    "user",
+		Content: input,
+	})
+
+	jsonRequest, err := json.Marshal(requestInfo)
+
+	if err != nil {
+		log.Fatal("Failed to build user request")
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonRequest))
+
+	if err != nil {
+		log.Fatal("Some error has occured.\nError:", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	if apiKey != "" {
+		req.Header.Set("Authorization", "Bearer "+apiKey)
+	}
+
+	return client.Do(req)
+}
+
+func GetMainText(line string) (mainText string) {
+	var obj = "{}"
+	if len(line) > 1 {
+		obj = strings.Split(line, "data: ")[1]
+	}
+
+	var d structs.CommonResponse
+	if err := json.Unmarshal([]byte(obj), &d); err != nil {
+		return ""
+	}
+
+	if len(d.Choices) > 0 {
+		mainText = d.Choices[0].Delta.Content
+		return mainText
+	}
+	return ""
+}

--- a/src/providers/opencode/opencode.go
+++ b/src/providers/opencode/opencode.go
@@ -95,7 +95,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/pollinations/pollinations.go
+++ b/src/providers/pollinations/pollinations.go
@@ -98,7 +98,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

--- a/src/providers/providers.go
+++ b/src/providers/providers.go
@@ -12,6 +12,7 @@ import (
 	"github.com/aandrew-me/tgpt/v2/src/providers/koboldai"
 	"github.com/aandrew-me/tgpt/v2/src/providers/minimax"
 	"github.com/aandrew-me/tgpt/v2/src/providers/ollama"
+	"github.com/aandrew-me/tgpt/v2/src/providers/opencode"
 	"github.com/aandrew-me/tgpt/v2/src/providers/openai"
 	"github.com/aandrew-me/tgpt/v2/src/providers/pollinations"
 	"github.com/aandrew-me/tgpt/v2/src/providers/powerbrain"
@@ -21,7 +22,7 @@ import (
 )
 
 var availableProviders = []string{
-	"", "deepseek", "isou", "gemini", "groq", "kimi", "koboldai", "minimax", "ollama", "openai", "pollinations", "powerbrain", "sky",
+	"", "deepseek", "isou", "gemini", "groq", "kimi", "koboldai", "minimax", "ollama", "opencode", "openai", "pollinations", "powerbrain", "sky",
 }
 
 func GetMainText(line string, provider string, input string) string {
@@ -42,6 +43,8 @@ func GetMainText(line string, provider string, input string) string {
 		return minimax.GetMainText(line)
 	case "ollama":
 		return ollama.GetMainText(line)
+	case "opencode":
+		return opencode.GetMainText(line)
 	case "openai":
 		return openai.GetMainText(line)
 	case "pollinations":
@@ -85,6 +88,8 @@ func NewRequest(input string, params structs.Params, extraOptions structs.ExtraO
 		return minimax.NewRequest(input, params)
 	case "ollama":
 		return ollama.NewRequest(input, params)
+	case "opencode":
+		return opencode.NewRequest(input, params)
 	case "openai":
 		return openai.NewRequest(input, params)
 	case "pollinations":

--- a/src/providers/sky/sky.go
+++ b/src/providers/sky/sky.go
@@ -71,7 +71,7 @@ func NewRequest(input string, params structs.Params) (*http.Response, error) {
 
 func GetMainText(line string) (mainText string) {
 	var obj = "{}"
-	if len(line) > 1 {
+	if strings.Contains(line, "data: ") {
 		obj = strings.Split(line, "data: ")[1]
 	}
 

